src/realm/mixed.cpp

```diff
@@ -85,88 +85,207 @@ inline int compare_float(Float a_raw, Float b_raw)
     // nans are treated as being less than all non-nan values
     return a_nan ? -1 : 1;
 }
-} // namespace _impl
 
-int Mixed::compare(const Mixed& b) const
+template <typename T>
+inline int compare_generic(T lhs, T rhs)
 {
-    // Comparing types first makes it possible to make a sort of a list of Mixed
-    // This will also handle the case where null values are considered lower than all other values
-    if (m_type > b.m_type)
+    return lhs == rhs ? 0 : lhs < rhs ? -1 : 1;
+}
+
+inline int compare_decimals(Decimal128 lhs, Decimal128 rhs)
+{
+    return lhs.compare(rhs);
+}
+
+inline int compare_decimal_to_double(Decimal128 lhs, double rhs)
+{
+    return lhs.compare(Decimal128(rhs)); // FIXME: slow and not accurate in all cases
+}
+
+// This is the tricky one. Needs to support the following cases:
+// * Doubles with a fractional component.
+// * Longs that can't be precisely represented as a double.
+// * Doubles outside of the range of Longs (including +/- Inf).
+// * NaN (defined by us as less than all Longs)
+// * Return value is always -1, 0, or 1 to ensure it is safe to negate.
+inline int compare_long_to_double(int64_t lhs, double rhs)
+{
+    // All Longs are > NaN
+    if (std::isnan(rhs))
         return 1;
-    else if (m_type < b.m_type)
-        return -1;
 
-    // Now we are sure the two types are the same
+    // Ints with magnitude <= 2**53 can be precisely represented as doubles.
+    // Additionally, doubles outside of this range can't have a fractional component.
+    static const int64_t kEndOfPreciseDoubles = 1ll << 53;
+    if (lhs <= kEndOfPreciseDoubles && lhs >= -kEndOfPreciseDoubles) {
+        return compare_float(double(lhs), rhs);
+    }
+
+    // Large magnitude doubles (including +/- Inf) are strictly > or < all Longs.
+    static const double kBoundOfLongRange = -static_cast<double>(LLONG_MIN); // positive 2**63
+    if (rhs >= kBoundOfLongRange)
+        return -1; // Can't be represented in a Long.
+    if (rhs < -kBoundOfLongRange)
+        return 1; // Can be represented in a Long.
+
+    // Remaining Doubles can have their integer component precisely represented as long longs.
+    // If they have a fractional component, they must be strictly > or < lhs even after
+    // truncation of the fractional component since low-magnitude lhs were handled above.
+    return compare_generic(lhs, int64_t(rhs));
+}
+} // namespace _impl
+
+bool Mixed::is_comparable(const Mixed& l, const Mixed& r)
```

Comment:
nit: lhs, rhs is used in the previous function, now it's l, r for the parameters. Later it's m1, m2. 
Perhaps use consistent parameter naming?

Reply:
Did some renaming